source/material_model/steinberger.cc

```diff
@@ -513,16 +532,15 @@ namespace aspect
           equation_of_state.initialize_simulator (this->get_simulator());
           equation_of_state.parse_parameters(prm);
 
-          // Check if compositional fields represent a composition
-          const std::vector<CompositionalFieldDescription> composition_descriptions = this->introspection().get_composition_descriptions();
-
           // Assign background field and do some error checking
-          has_background_field = (equation_of_state.number_of_lookups() == this->introspection().n_chemical_composition_fields() + 1);
+          has_background_field = ((equation_of_state.number_of_lookups() == 1) ||
```

Comment:
@jdannberg I've decided to simplify the logic in the source code and require that the type is specified for fields that aren't of type "chemical composition".

Reply:
I am not really happy with this solution. It is not just because of not being backwards compatible, but it would also be inconsistent with all of our other parameters (which @gassmoeller pointed out when we discussed this). For all other parameters, you can just specify one value, and this value will be used for all fields (of whatever field type). So you should be able to just specify one table, which will then be used for all fields (independent of field type). For tables this is actually even more important than for other parameters, because it may require a significant amount of memory to read in the same table several times. 

Is there no way to do this right with the original logic of having a background field?

Reply:
Ok, first let me describe what I think `steinberger_several_fields.prm` should do:

- There are three compositional fields and one lookup.
- The compositional fields *don't* correspond to chemical compositions (values from 0 to 3e7). They represent some other implicitly tracked variable.
- The field type for these fields is "chemical composition" (the default value), but on main branch this is ignored in `steinberger.cc` because there is just one lookup.

Now I'll describe what I originally implemented (https://github.com/geodynamics/aspect/pull/5362/files/06ce80544b7e64ea1025552d25178cf202f77b03) and how that maintained the status-quo:
- On main branch, there was a background field if there was one look-up and no compositional fields, but not if there was one look-up and some compositional fields.
- In this PR, I changed that so that there is always a background field if there is one look-up. That was in better agreement with the descriptions in the two tests: `It is intended to test the ability of the steinberger material model *to use the first data file as a background composition*, and ignore other compositional fields if only a single material file is provided.`
- This didn't change any existing functionality, but it does mean that users cannot use those fields to modify viscosities.

To answer your questions:
> If I understand the code below correctly, in this case you would erase the name of this field from the list of allowed keys, so if one tried to use the name of this field to specify the viscosity, it would not work.

Yes, that is (I think) the correct behaviour according to the test descriptions, which suggest that if the is one lookup, compositional fields don't correspond to different material proportions. 

> Also what happens if there are 5 chemical fields and one look-up? Do we want to allow the user to specify 5 different compositional prefactors in that case? 

No, that would be inconsistent with the current test descriptions, and it would also lead to undefined behaviour. How would we know which compositional fields correspond to chemical compositions (given that the behaviour on main branch explicitly ignores fields of type `chemical composition`)?

> So you should be able to just specify one table, which will then be used for all fields (independent of field type). For tables this is actually even more important than for other parameters, because it may require a significant amount of memory to read in the same table several times.

On main branch, if there's one table, all compositional fields are explicitly ignored by the material model.

Reply:
I agree that the solution I implemented after your initial comments is further away from what you envisage. Sorry, I misunderstood your message. I'm happy to remove that commit - we could probably merge the PR then and discuss further changes separately.

One way to implement what you have in mind (one lookup, multiple viscosity prefactors) would be to add a further parameter cross-referencing the lookups to the fields of type "chemical composition" (i.e. a vector of unsigned ints or strings corresponding to the field names). I think the number of viscosity prefactors (after parsing) should always be equal to the number of fields of type chemical composition, so `steinberger_phases.prm` and `steinberger_several_fields.prm` would still need to be modified to specify the field types, and that would still break backwards compatibility.

I can't think of another way to do what you want without serious opportunity for confusion, so I'd appreciate any ideas you might have!

Reply:
I think allowing different viscosity prefactors if there is only one look-up is a separate issue we don’t have to discuss now, since adding functionality later in another PR is always fine, so I totally agree with you on that. (And I can think about ideas for how to do that.)

If everything works as before when you just remove the “remove one lookup case from steinberger“ commit, I think that would be a good solution. 

That still leaves the change with the name of the background field, i.e., if it is called background or whatever name the user gave it. I think it is okay to change the name of the key from how it is now on main (name of the compositional field) to background, and I agree that is more consistent. The case where it would break input files is so special that I just hope it won't be an issue. 

The part that was important to me is keeping the option to have several fields (compositional fields, chemical fields, whatever) and to just read in one table for all of them. And from your second message, it seems like that would work when removing the “remove one lookup case from steinberger“ commit?

Just to explain, this is intended for the material model, since the documentation of the “Material file names” parameter says

> The file names of the material data (material data is assumed to be in order with the ordering of the compositional fields). Note that there are three options on how many files need to be listed here: 1. *If only one file is provided, it is used for the whole model domain, and compositional fields are ignored.* 2. If there is one more file name than the number of compositional fields, then the first file is assumed to define a `background composition' that is modified by the compositional fields. If there are exactly as many files as compositional fields, the fields are assumed to represent the fractions of different materials and the average property is computed as a sum of the value of the compositional field times the material property of that field.